Reduce granularity of ServiceWorker process assertions/activities

youennf opened the pull request:
#### 0218ef6e826361d6ed71581b980887d4925a0ade
<pre>
Reduce granularity of ServiceWorker process assertions/activities
<a href="https://rdar.apple.com/137705981">rdar://137705981</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=282828">https://bugs.webkit.org/show_bug.cgi?id=282828</a>

Reviewed by Ben Nham.

Before the patch, we were taking service worker process assertion based on the clients of the service worker.
If clients had forward or backward process assertions, we would take the same assertion for the service worker process.
The exception was for functional events where we could run service workers without clients.

We are now reusing the same principle as functional events by having network process monitor whether a service worker process should take a process assertion.
This is done by WebSWServerToContextConnection, which sends IPC messages to NetworkProcessProxy when there is a need to take or drop the process assertion.

WebSWServerToContextConnection decides whether taking a process assertion  by being notified each time one of its service worker needs to run.
This happens for all events sent to the service worker.

WebSWServerToContextConnection also makes sure to keep the process assertion as long as these events are not fully handled by the service worker.
This is done by checking the ServiceWorker can be idle.

We also add a check for ongoing fetches, since a service worker load might take time which are not captured by events.
This is implemented in WebSWServerToContextConnection::areServiceWorkersIdle.

When a WebSWServerToContextConnection takes an assertion, it will schedule a timer to try removing the process assertion, timer default value is 10 seconds.

We update WebProcessProxy::updateRemoteWorkerProcessAssertion to not take any process assertion when a service worker process is not doing any processing.
If the service worker process is doing some processing, we will either take a foreground or backgorund assertion based on the service worker process clients.

We move WebProcessProxy::m_hasServiceWorkerBackgroundProcessing to WebProcessProxy::m_serviceWorkerInformation::hasBackgroundProcessing.
This allows to set it back this flag to false by nullifying WebProcessProxy::m_serviceWorkerInformation in disableRemoteWorkers.
We can also remove the call to endServiceWorkerBackgroundProcessing in NetworkProcessProxy::processHasUnresponseServiceWorker, since this will be handled in disableRemoteWorkers.

* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::runServiceWorkerIfNecessary):
(WebCore::SWServer::areServiceWorkersIdle):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::WebSWServerToContextConnection):
(WebKit::WebSWServerToContextConnection::installServiceWorkerContext):
(WebKit::WebSWServerToContextConnection::fireInstallEvent):
(WebKit::WebSWServerToContextConnection::fireActivateEvent):
(WebKit::WebSWServerToContextConnection::firePushEvent):
(WebKit::WebSWServerToContextConnection::fireNotificationEvent):
(WebKit::WebSWServerToContextConnection::fireBackgroundFetchEvent):
(WebKit::WebSWServerToContextConnection::fireBackgroundFetchClickEvent):
(WebKit::WebSWServerToContextConnection::terminateWorker):
(WebKit::WebSWServerToContextConnection::workerTerminated):
(WebKit::WebSWServerToContextConnection::startProcessAssertionTimer):
(WebKit::WebSWServerToContextConnection::serviceWorkerNeedsRunning):
(WebKit::WebSWServerToContextConnection::areServiceWorkersIdle const):
(WebKit::WebSWServerToContextConnection::processAssertionTimerFired):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processHasUnresponseServiceWorker):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::updateServiceWorkerProcessAssertion):
(WebKit::WebProcessProxy::updateSharedWorkerProcessAssertion):
(WebKit::WebProcessProxy::updateRemoteWorkerProcessAssertion):
(WebKit::WebProcessProxy::startServiceWorkerBackgroundProcessing):
(WebKit::WebProcessProxy::endServiceWorkerBackgroundProcessing):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:

Canonical link: <a href="https://commits.webkit.org/286796@main">https://commits.webkit.org/286796@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79d7ba0aea1311c9d441c54b4394802adfeb24ee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76454 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55489 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29360 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80979 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27730 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78571 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64631 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3782 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59950 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18062 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79521 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49861 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65653 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40274 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47256 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23142 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26053 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68382 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23474 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82427 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3830 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2517 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68224 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3983 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65622 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67469 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16950 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11441 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9534 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3777 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6586 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3800 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7230 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5558 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
```diff
@@ -467,5 +439,39 @@ void WebSWServerToContextConnection::setInspectable(ServiceWorkerIsInspectable i
+        return;
+    }
+
+    m_isTakingProcessAssertion = false;
```
What happens if this object in NetworkProcess is destructed before the `EndServiceWorkerBackgroundProcessing` IPC can be sent to the UIProcess?

I tried to somewhat address this issue [286417@main](https://commits.webkit.org/286417@main) in by adding a call to `endServiceWorkerBackgroundProcessing()` in `processHasUnresponseServiceWorker()`, but maybe we need something more generic?

Similarly, I guess we need to make sure the activity is cleaned up in UIProcess if NetworkProcess dies before the `EndServiceWorkerBackgroundProcessing` IPC can be sent.

youennf replied:
In case of network process crash, we terminate all service workers, which calls `disableRemoteWorkers`, which will destroy all process assertions. So we should be good.

There is an edge case we are not covering, which is `WebProcessProxy::m_hasServiceWorkerBackgroundProcessing`.
It could be kept to `true` which could be an issue if the process is again used for service workers.

I'll move WebProcessProxy::m_hasServiceWorkerBackgroundProcessing to WebProcessProxy::m_serviceWorkerInformation::hasBackgroundProcessing.
This way, `disableRemoteWorkers` will reset this flag as well.

bnham reviewed (commented):
Seems reasonable at first glance. Can you let me know if the concerns about activity cleanup in the case of NetworkProcess crash or if the connection is destroyed are important (e.g. due to service worker termination)?

nit: `If clients had forward or backward process assertions` should be `If clients had foreground or background process assertions`

bnham reviewed (approved):
LGTM

youennf committed:
Reduce granularity of ServiceWorker process assertions/activities
rdar://137705981
https://bugs.webkit.org/show_bug.cgi?id=282828

Reviewed by Ben Nham.

Before the patch, we were taking service worker process assertion based on the clients of the service worker.
If clients had forward or backward process assertions, we would take the same assertion for the service worker process.
The exception was for functional events where we could run service workers without clients.

We are now reusing the same principle as functional events by having network process monitor whether a service worker process should take a process assertion.
This is done by WebSWServerToContextConnection, which sends IPC messages to NetworkProcessProxy when there is a need to take or drop the process assertion.

WebSWServerToContextConnection decides whether taking a process assertion  by being notified each time one of its service worker needs to run.
This happens for all events sent to the service worker.

WebSWServerToContextConnection also makes sure to keep the process assertion as long as these events are not fully handled by the service worker.
This is done by checking the ServiceWorker can be idle.

We also add a check for ongoing fetches, since a service worker load might take time which are not captured by events.
This is implemented in WebSWServerToContextConnection::areServiceWorkersIdle.

When a WebSWServerToContextConnection takes an assertion, it will schedule a timer to try removing the process assertion, timer default value is 10 seconds.

We update WebProcessProxy::updateRemoteWorkerProcessAssertion to not take any process assertion when a service worker process is not doing any processing.
If the service worker process is doing some processing, we will either take a foreground or backgorund assertion based on the service worker process clients.

We move WebProcessProxy::m_hasServiceWorkerBackgroundProcessing to WebProcessProxy::m_serviceWorkerInformation::hasBackgroundProcessing.
This allows to set it back this flag to false by nullifying WebProcessProxy::m_serviceWorkerInformation in disableRemoteWorkers.
We can also remove the call to endServiceWorkerBackgroundProcessing in NetworkProcessProxy::processHasUnresponseServiceWorker, since this will be handled in disableRemoteWorkers.

* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::runServiceWorkerIfNecessary):
(WebCore::SWServer::areServiceWorkersIdle):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::WebSWServerToContextConnection):
(WebKit::WebSWServerToContextConnection::installServiceWorkerContext):
(WebKit::WebSWServerToContextConnection::fireInstallEvent):
(WebKit::WebSWServerToContextConnection::fireActivateEvent):
(WebKit::WebSWServerToContextConnection::firePushEvent):
(WebKit::WebSWServerToContextConnection::fireNotificationEvent):
(WebKit::WebSWServerToContextConnection::fireBackgroundFetchEvent):
(WebKit::WebSWServerToContextConnection::fireBackgroundFetchClickEvent):
(WebKit::WebSWServerToContextConnection::terminateWorker):
(WebKit::WebSWServerToContextConnection::workerTerminated):
(WebKit::WebSWServerToContextConnection::startProcessAssertionTimer):
(WebKit::WebSWServerToContextConnection::serviceWorkerNeedsRunning):
(WebKit::WebSWServerToContextConnection::areServiceWorkersIdle const):
(WebKit::WebSWServerToContextConnection::processAssertionTimerFired):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processHasUnresponseServiceWorker):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::updateServiceWorkerProcessAssertion):
(WebKit::WebProcessProxy::updateSharedWorkerProcessAssertion):
(WebKit::WebProcessProxy::updateRemoteWorkerProcessAssertion):
(WebKit::WebProcessProxy::startServiceWorkerBackgroundProcessing):
(WebKit::WebProcessProxy::endServiceWorkerBackgroundProcessing):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ServiceWorkerBasic.mm:

Canonical link: https://commits.webkit.org/286796@main
